website/blog/2019-08-19-tablescan-structs.md

```diff
@@ -0,0 +1,163 @@
+---
+title: Table Scan: Doing The Right Thing With Structured Types
+author: Orri Erling
+authorURL: https://www.linkedin.com/in/orrierling/
+authorFBID: 100026224749124
+---
+
+In the previous article we saw what gains are possible when filtering early and in the right order. In this article we look at how we do this with nested and structured types.
+
+<!--truncate-->
+
+We use the 100G TPC-H dataset but now we group top level columns into structs or maps. 
+
+The tables are defined as follows:
+
+```sql
+CREATE TABLE hive.tpch.exportlineitem  (
+    l_orderkey BIGINT,
+    l_linenumber INTEGER,
+    l_shipment row(
+        l_partkey BIGINT,
+        l_suppkey BIGINT,
+        l_extendedprice DOUBLE,
+        l_discount DOUBLE,
+        l_quantity DOUBLE,
+        l_shipdate DATE,
+        l_receiptdate DATE,
+        l_commitdate DATE,
+        l_comment VARCHAR(44)),
+    l_export row(
+        s_nation BIGINT,
+        c_nation BIGINT,
+        is_inside_eu INTEGER,
+        is_restricted INTEGER,
+        license row(
+            applydate DATE,
+            grantdate DATE,
+            filing_no BIGINT,
+            comment VARCHAR))
+)
+WITH (
+    format = 'ORC'
+);
+```
+
+The l_shipment struct has the non-null top-level columns we all have come to know and love. The l_export struct is null if the customer and supplier nations are the same and present otherwise. A fraction of the rows have an additional nested export license struct.
+
+```sql
+CREATE TABLE hive.tpch.lineitem_map (
+    l_orderkey BIGINT,
+    l_linenumber INTEGER,
+    ints map(INTEGER, BIGINT),
+    strs map(VARCHAR(2), varchar(44))
```

Comment:
- varchar -> VARCHAR
- is the max size of the varchars important here?

Reply:
Fixed and removed sizes.